designs/dynamic-bundle.md

```diff
@@ -0,0 +1,267 @@
+Dynamic Bundle
+===================
+
+## Overview
+Kubebuilder currently has an implementation of a `Bundle` which allows multiple plugins to be *bundled* together. The execution of the plugins within a `Bundle` follows a sequential order based on their definition. 
+
+An example of the current method:
+
+```go
+gov3Bundle, _ := plugin.NewBundle(golang.DefaultNameQualifier, plugin.Version{Number: 3},
+	kustomizecommonv1.Plugin{},
+	golangv3.Plugin{},
+)
+```
+
+In the above example a plugin `Bundle` is created that ensures that the `kustomize/v1` plugin is run before the `go/v3` plugin.
+ 
+With the addition of Phase 2 Plugins on the way, users using the `--plugins` flag to utilize external plugins will not get the benefits of using some default functionality that previous Kubebuilder users using the default `go/v3` plugin gained via the `kustomize/v1` plugin.
+
+## Proposal
+The proposed solution is to create a new form of `Bundle` that is more dynamic and can be used to create a default bundle that contains sensible default plugins that can be used with all plugins. This could be overriden by a command line flag.
```

Comment:
Would like to understand more about the use-cases of the `DynamicBundle`.

Currently, in Kubebuilder, there are `PreScaffold` and `PostScaffold` hooks which can be used by every subcommand (init, create api, create webhook). Pre scaffold logic allows to ensure the correct go version, validating that main.go exists, so on. Similarly, post scaffold logic can execute a set of tasks after the main scaffolding is done, such as updating dependencies.  I'm wondering if these can be customized further to set the sane defaults during `PostScaffold` logic, without having to introduce new default plugins.

Reply:
I think the biggest use case is to make it so that we don't have to create a new `Bundle` every time we create a new language plugin.

The sane defaults are more so default plugins that should be run. Hopefully this comment should clear it up a bit more but let me know if you need more details: https://github.com/kubernetes-sigs/kubebuilder/pull/2671#discussion_r865863397